### ISSUE TITLE
Warning with `use client` directive with Vite

### ISSUE THREAD

**Issue by hckhanh** [TanStack/query#5175]
### Describe the bug

Module level directives cause errors when bundled, "use client" in "node_modules/@tanstack/react-query/build/lib/useSyncExternalStore.mjs" was ignored.

### Your minimal, reproducible example

I don't have time

### Steps to reproduce

Use query v4.28.0 with vite. I wonder does it affect the build or not.

### Expected behavior

I will not see the warning

### How often does this bug happen?

Every time

### Screenshots or Videos

![image](https://user-images.githubusercontent.com/6380436/226784130-8d46f637-ce62-4717-b19a-55110cf6ee65.png)

### Platform

Windows: 11

### Tanstack Query adapter

react-query

### TanStack Query version

v4.28.0

### TypeScript version

v5

**Comment by TkDodo** [comment]
please see the discussion on the PR:

- https://github.com/TanStack/query/pull/5161#issuecomment-1475871864

@Ephem I don't think there's anything we can do right now about this?

**Comment by Ephem** [comment]
No, besides removing the directives entirely there's nothing we can do in React Query itself sadly, at least that I know of. A bit swamped this week but I'll try to find some time to open issues (if none exist) and perhaps contribute an ignore for the warning to the Vite React plugin.

TL;DR for this warning: You can safely ignore it. If you want to hide it, you can use Rollups ` onwarn` handler. Hopefully hiding the warning manually wont be necessary in the future. Details in comments in the PR.

**Comment by rodolfocerezo** [comment]
Hi, I'm also getting the same errors, at least we know now that is safe to ignore them.

```
vite v4.2.1 building for production...
Module level directives cause errors when bundled, "use client" in "node_modules/@tanstack/react-query-devtools/build/lib/index.mjs" was ignored.
Module level directives cause errors when bundled, "use client" in "node_modules/@tanstack/react-query/build/lib/useQueries.mjs" was ignored.
Module level directives cause errors when bundled, "use client" in "node_modules/@tanstack/react-query/build/lib/useQuery.mjs" was ignored.
Module level directives cause errors when bundled, "use client" in "node_modules/@tanstack/react-query/build/lib/QueryClientProvider.mjs" was ignored.
Module level directives cause errors when bundled, "use client" in "node_modules/@tanstack/react-query/build/lib/Hydrate.mjs" was ignored.
Module level directives cause errors when bundled, "use client" in "node_modules/@tanstack/react-query/build/lib/QueryErrorResetBoundary.mjs" was ignored.
Module level directives cause errors when bundled, "use client" in "node_modules/@tanstack/react-query/build/lib/useIsFetching.mjs" was ignored.
Module level directives cause errors when bundled, "use client" in "node_modules/@tanstack/react-query/build/lib/useIsMutating.mjs" was ignored.
Module level directives cause errors when bundled, "use client" in "node_modules/@tanstack/react-query/build/lib/useInfiniteQuery.mjs" was ignored.
Module level directives cause errors when bundled, "use client" in "node_modules/@tanstack/react-query/build/lib/useMutation.mjs" was ignored.
Module level directives cause errors when bundled, "use client" in "node_modules/@tanstack/react-query/build/lib/isRestoring.mjs" was ignored.
Module level directives cause errors when bundled, "use client" in "node_modules/@tanstack/react-query-devtools/build/lib/devtools.mjs" was ignored.
Module level directives cause errors when bundled, "use client" in "node_modules/@tanstack/react-query/build/lib/reactBatchedUpdates.mjs" was ignored.
Module level directives cause errors when bundled, "use client" in "node_modules/@tanstack/react-query/build/lib/useSyncExternalStore.mjs" was ignored.
Module level directives cause errors when bundled, "use client" in "node_modules/@tanstack/react-query/build/lib/errorBoundaryUtils.mjs" was ignored.
Module level directives cause errors when bundled, "use client" in "node_modules/@tanstack/react-query/build/lib/useBaseQuery.mjs" was ignored.
Module level directives cause errors when bundled, "use client" in "node_modules/@tanstack/react-query-devtools/build/lib/theme.mjs" was ignored.
Module level directives cause errors when bundled, "use client" in "node_modules/@tanstack/react-query-devtools/build/lib/Explorer.mjs" was ignored.
```

**Comment by rodolfocerezo** [comment]
As @Ephem said, we can add the following in our rollup options:

```ts
onwarn(warning, warn) {
  if (warning.code === 'MODULE_LEVEL_DIRECTIVE') {
    return
  }
  warn(warning)
}
```

**Comment by duypham90** [comment]
I got many error warning when build production with React + Vite v4.2.1
Not use NextJS

<img width="1633" alt="Screen Shot 2023-03-31 at 21 09 59" src="https://user-images.githubusercontent.com/24804332/229144080-0c9dc152-4d25-499f-894f-d52c113a109c.png">

**Comment by Ephem** [comment]
As an update to anyone reading this, the Vite React-plugin will soon silence this warning by default. This is not available in a stable release yet, but is in the [4.0.0-beta.0 release](https://github.com/vitejs/vite-plugin-react/releases/tag/plugin-react%404.0.0-beta.0).

Thanks a lot to @ArnaudBarre for adding the fix and @thebuilder for reporting it over there! 👏

**Comment by thebuilder** [comment]
Sounds good 👌

**Comment by derekdavenport** [comment]
> As @Ephem said, we can add the following in our rollup options:
> 
> ```ts
> onwarn(warning, warn) {
>   if (warning.code === 'MODULE_LEVEL_DIRECTIVE') {
>     return
>   }
>   warn(warning)
> }
> ```

This doesn't work when I use the nodePolyfills plugin. Why would that be?

I can't build on my server because the heap runs out of memory from all these warnings.

**Comment by nemonemi** [comment]
I can still see it with version 4.1.0.

**Comment by ljmerza** [comment]
im not even using react-query and i get this error on latest vite

**Comment by noherczeg** [comment]
I'm using Vite `4.5.0` + @vitejs/plugin-react `4.1.0` and the warnings are gone.

**Comment by KholdStare** [comment]
@derekdavenport I'm still seeing the same issue with node polyfill plugin, with latest versions of everything: https://github.com/davidmyersdev/vite-plugin-node-polyfills/issues/49

**Comment by bogdancss** [comment]
`vite`: `4.5.0`
`@vitejs/plugin-react`: `4.1.0` 
`vite-plugin-node-polyfills`: `0.15.0`

We are still getting the warnings with these versions

**Comment by ArnaudBarre** [comment]
Yeah this is coming from `vite-plugin-node-polyfills`, I sent a fix: https://github.com/davidmyersdev/vite-plugin-node-polyfills/pull/53

**Comment by MuhammadHasher017** [comment]
> > As @Ephem said, we can add the following in our rollup options:
> > ```ts
> > onwarn(warning, warn) {
> >   if (warning.code === 'MODULE_LEVEL_DIRECTIVE') {
> >     return
> >   }
> >   warn(warning)
> > }
> > ```
> 
> Which file am I going to add this?

```

export default defineConfig({
  build: {
    chunkSizeWarningLimit: 100,
    rollupOptions: {
      onwarn(warning, warn) {
        if (warning.code === "MODULE_LEVEL_DIRECTIVE") {
          return;
        }
        warn(warning);
      },
    },
  },
  plugins: [react()],
  resolve: {
    alias: {
      "@": "/src",
    },
  },
});




```
You can add it in vite.config.js like that.

**Comment by sean080597** [comment]
I resolved the problem
1) Updating these versions in package.json:
"@vitejs/plugin-react": "^4.3.2"
"laravel-vite-plugin": "^1.0.5",
"vite": "^5.4.9",
"vite-plugin-node-polyfills": "^0.22.0"
2) Ensure your package.json contains "type": "module".

**Comment by mariusbutuc** [comment]
> As [@Ephem](https://github.com/Ephem) said, we can add the following in our rollup options:
> ```js
> onwarn(warning, warn) {
>   if (warning.code === 'MODULE_LEVEL_DIRECTIVE') {
>     return
>   }
>   warn(warning)
> }
> ```

For some reason, using Rollup's `onwarn` handler didn't have the expected outcome in our case. Looking at [Vite's docs on customizing the build](https://v3.vitejs.dev/guide/build.html#customizing-the-build) and the underlying Rollup options, we got the expected outcome by adapting the example to use [the `onLog` handler](https://rollupjs.org/configuration-options/#onlog) instead:

```js
onLog(level, log, handler) {
  if (log.code === "MODULE_LEVEL_DIRECTIVE") {
    return;
  }
  handler(level, log);
},
```

**Comment by Onlylonger** [comment]
Also can reference this [comment](https://github.com/vitejs/vite/issues/15012#issuecomment-1825035992). A summary rather than a line-by-line display might be more appropriate

**_mute-warnings-plugin.ts_**
```ts
import { Plugin } from "vite";

export const muteWarningsPlugin = (warningCodes: string[]): Plugin => {
  const mutedMessages = new Map();

  return {
    name: "mute-warnings",
    enforce: "pre",
    config: () => ({
      build: {
        rollupOptions: {
          onwarn(warning, defaultHandler) {
            const code = warning.code;

            if (code) {
              if (warningCodes.includes(code) && !mutedMessages.has(code)) {
                mutedMessages.set(code, {
                  exMsg: warning.message,
                });
              }

              if (mutedMessages.has(code)) {
                return;
              }
            }

            defaultHandler(warning);
          },
        },
      },
    }),
    closeBundle() {
      this.warn(
        "Some of your muted warnings never appeared during the build process:"
      );
      for (const [key, item] of mutedMessages) {
        this.warn(`\t(${key} example): ${item.exMsg}`);
      }
    },
  };
};

```

**_vite.config.ts_**
```ts
export default defineConfig(({ mode }) => ({
  plugins: [
   ...
    muteWarningsPlugin([
      "SOURCEMAP_ERROR",
      "MODULE_LEVEL_DIRECTIVE",
    ]),
  ]
})
```
